src/main/java/net/mcreator/blockly/java/ProcedureCodeOptimizer.java

```diff
@@ -0,0 +1,115 @@
+/*
+ * MCreator (https://mcreator.net/)
+ * Copyright (C) 2012-2020, Pylo
+ * Copyright (C) 2020-2021, Pylo, opensource contributors
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+package net.mcreator.blockly.java;
+
+@SuppressWarnings("unused") public class ProcedureCodeOptimizer {
+	enum ParseState {
+		INSIDE_INLINE_COMMENT, INSIDE_COMMENT_BLOCK, INSIDE_STRING, INSIDE_STRING_ESCAPE_SEQUENCE, OUTSIDE
+	}
+
+	/**
+	 * This method attempts to remove the parentheses surrounding the given code, if they are paired.
+	 * @param code The code to optimize
+	 * @return If possible, the code without surrounding parentheses
+	 */
+	public static String removeParentheses(String code) {
+		String toClean = code.strip();
+		if (toClean.startsWith("(") && toClean.endsWith(")")) {
+			var state = ParseState.OUTSIDE;
+			int parentheses = 1;
+			char prevChar = '(';
+			int backSlashCounter = 0;
+			for (int i = 1; i < toClean.length() - 1; i++) {
+				char c = toClean.charAt(i);
+				switch (state) {
+				case OUTSIDE:
+					if (c == '/' && prevChar == '/')
+						state = ParseState.INSIDE_INLINE_COMMENT;
+					else if (c == '*' && prevChar == '/')
+						state = ParseState.INSIDE_COMMENT_BLOCK;
+					else if (c == '"')
+						state = ParseState.INSIDE_STRING;
+					else if (c == '(')
+						parentheses++;
+					else if (c == ')' && --parentheses == 0) // The first ( isn't paired with the last ), we can't remove them
+						return code;
+					break;
+				case INSIDE_INLINE_COMMENT:
+					if (c == '\n' || c == '\r')
+						state = ParseState.OUTSIDE;
+					break;
+				case INSIDE_STRING_ESCAPE_SEQUENCE:
+					if (c == '\\') {
+						backSlashCounter++;
+						break;
+					}
+					state = ParseState.INSIDE_STRING;
+				case INSIDE_STRING:
+					if (c == '\\') {
+						state = ParseState.INSIDE_STRING_ESCAPE_SEQUENCE;
+						backSlashCounter = 0;
+					} else if (c == '"' && (prevChar != '\\' || backSlashCounter % 2 != 0)) {
+						state = ParseState.OUTSIDE;
+					}
+					break;
+				case INSIDE_COMMENT_BLOCK:
+					if (c == '/' && prevChar == '*')
+						state = ParseState.OUTSIDE;
+					break;
+				}
+				prevChar = c;
+			}
+			if (state == ParseState.OUTSIDE && --parentheses == 0) // The last character is a valid ")"

```

Comment:
I am really sorry I am asking that much :D, but are there any actual cases where the parentheses are not matched and can't be removed here?

Reply:
The last ) could be in an inline comment

Reply:
Optimizer runs for procedures that are "output" by nature, eg.

`(something.getData())`

and not on statements like

`setData(something); // another (`

if I am not mistaken.

So procedure like

`(something.getData()) // another (`

would anyway break the pipeline down the stream, eg. when inserted into

`setData((something.getData()) // another (); // another (`

so inline comment cases should not be possible.

Or am I missing something?

Reply:
That's true, I removed the check

Reply:
But is in this case the parentheses matching even needed, if the procedure starts and ends with (), then it is safe to remove the parentheses under the assumption above. Or not?

Reply:
Cases like `(something).getData()`, where it starts and ends with () but can't remove those, are already caught by a check inside the loop, so we can assume the parentheses are paired if we got out of the loop

Reply:
Oh you are right, I totally missed cases like that :D